Metagenome Analysis at [STAMPS 2017](https://stamps.mbl.edu/index.php/Schedule)
===============================================================================

Instructor: [Tandy Warnow](http://tandy.cs.illinois.edu)  
Teaching Assistant: [Erin Molloy](http://emolloy2.web.engr.illinois.edu)

This is a tutorial for the **SEPP** and **TIPP** software packages developed in the Warnow Lab by [Siavash Mirarab](http://eceweb.ucsd.edu/~smirarab/) and [Nam Nguyen](https://sites.google.com/a/eng.ucsd.edu/namphuon/home). To begin the hands-on part of the tutorial on the MBL servers, clone this repository in your home directory,
```
git clone https://github.com/ekmolloy/stamps-tutorial.git
```

load the python module,
```
module purge
module load python/2.7.12-201701011205
```
and follow the [tutorial worksheet](tutorial.md). Please flag us down when you have questions, we are happy to help! 

This tutorial has been set-up to run on the MBL servers. To use SEPP and TIPP in the future, you may need to
+ Download the SEPP/TIPP software package [here](https://github.com/smirarab/sepp).
+ Work through additional tutorials [here](https://github.com/smirarab/sepp/tree/master/tutorial).

Citations
---------
Mirarab, Nguyen, and Warnow. (2012). SEPP: SATe-Enabled Phylogenetic Placement. *Proceedings of the 2012 Pacific Symposium on Biocomputing (PSB 2012)* 17:247-258. doi:[10.1093/bioinformatics/btu721](https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btu721)

Nguyen, Mirarab, Liu, Pop, and Warnow (2014). TIPP:Taxonomic Identification and Phylogenetic Profiling. *Bioinformatics* 30(24):3548-3555. doi:[10.1142/9789814366496_0024](http://www.worldscientific.com/doi/abs/10.1142/9789814366496_0024)
